File: 0238-product-of-array-except-self/0238-product-of-array-except-self.ts
function productExceptSelf(nums: number[]): number[] {
    const n = nums.length;
    const prefix: number[] = new Array(n).fill(1);
    const postfix: number[] = new Array(n).fill(1);
    const result: number[] = new Array(n);

    for (let i = 1; i < n; i++) {
        prefix[i] = nums[i - 1] * prefix[i - 1];
    }

    for (let i = n - 2; i >= 0; i--) {
        postfix[i] = nums[i + 1] * postfix[i + 1];
    }

    console.log(postfix);

    for (let i = 0; i < n; i++) {
        result[i] = prefix[i] * postfix[i];
    }

    return result;
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int pr[] = new int[nums.length];
        int su[] = new int[nums.length];
        pr[0] = 1;
        for (int i = 1; i < nums.length; i++) {
            pr[i] = pr[i - 1] * nums[i - 1];
        }
        su[nums.length - 1] = 1;
        for (int i = nums.length - 2; i >= 0; i--) {
            su[i] = su[i + 1] * nums[i + 1];
        }

        int res[] = new int[nums.length];
        for (int i = 0; i < nums.length; i++) {
            res[i] = pr[i] * su[i];
        }
        return res;
    }
}